Accelerated motif search algorithms on dna sequences

quim6 opened the pull request:
<!--- Please read each of the following items and confirm by replacing
 !--the [ ] with a [X] --->

- [X] I hereby agree to dual licence this and any previous contributions under both
the _Biopython License Agreement_ **AND** the _BSD 3-Clause License_.

- [X] I have read the ``CONTRIBUTING.rst`` file, have run ``pre-commit``
locally, and understand that continuous integration checks will be used to
confirm the Biopython unit tests and style checks pass with these changes.

- [X] I have added my name to the alphabetical contributors listings in the files
``NEWS.rst`` and ``CONTRIB.rst`` as part of this pull request, am listed
already, or do not wish to be listed. (*This acknowledgement is optional.*)

We have developed a set of functions for fast search of motif instances in DNA sequences. Specifically, we have implemented versions of the lookahead, permuted lookahead and super-alphabet algorithms, as described in: 

**Cinzia Pizzi, Esko Ukkonen** - [Fast profile matching algorithms — A survey](https://doi.org/10.1016/j.tcs.2008.01.015), *Theoretical Computer Science*, Volume 395, Issues 2–3, 2008, 137-157. ISSN 0304-3975.

These methods have been implemented within the Motifs package, as C functions for fast search with a Python wrapper, following the template of the existing `_pwm_calculate` function. These functions implement threshold-based search (returning only hits above a pre-specified threshold, as the existing `motifs.pssm.search` method). In contrast to that method, however, they do not call the C `calculate` function. Rather, the speedup (around 2x to 6x) provided by these methods is based on the ability to not compute all the scores in the sequence, dropping any sequences that cannot reach the threshold.

The Python `fast_search` wrapper function has been designed so that new methods for fast search can be efficiently incorporated into the extension.

funcylambda committed:
Updated docstrings to include the output and mention any validation that occurs. Also, changed docstring so they have consistent capitalization. Baseml >> BASEML etc

funcylambda committed:
Updated docstrings to use consistent capitalization. paml >> PAML etc.

funcylambda committed:
Update CONTRIB.rst

Commit:
Fixed flake8 rule D209 violations

D209 Multi-line docstring closing quotes should be on a separate line

rob-miller committed:
Add OpenSCAD code to enable multimaterial printer output

rob-miller committed:
make spheres nicer on first start

rob-miller committed:
update NEWS file

mdehoon commented on setup.py:
```diff
@@ -194,6 +195,15 @@ def can_import(module_name):
+            "Bio/motifs/search_algorithms.c",
+            "Bio/motifs/darray.c",
+        ],
+        include_dirs=[numpy.get_include()],
```
Do you need to include the numpy headers?

quim6 replied:
It's currently used to create and return a numpy array from the C function. 'py_search'
Is including NumPy headers problematic? If it is I can move the array creation to pure Python.

mdehoon replied:
> Is including NumPy headers problematic?

Yes, as the user may have not installed them. It doesn't matter with a binary distribution, but if they are installing from source, it tends to cause confusion.

> If it is I can move the array creation to pure Python.

Yes, this is the approach we take in other parts of Biopython. Figure out what the dimensions of the output array will be, create it in Python, and pass it to your C function. Then use the buffer interface to fill the array with the values calculated by the C function.

mdehoon replied:
Another solution is to define a new class corresponding to one `Hit`. That may be better anyway, because then you can store the position as an integer instead of a floating point value.

Or you can return two arrays, one for the positions and one for the scores.

quim6 commented on Bio/motifs/search_algorithms.c:
```diff
@@ -0,0 +1,383 @@
+*           - If the partial score is less than min_scores[j], break early (prune)
+*       - If the full subsequence score ≥ threshold, store the match.
+*
+* Argsuments:
```
Thanks for catching this! I'll fix the typo immediately and remove the unrelated style changes (caused by my Black formatting issues).
Sorry about the noise I will push a clean version shortly.

quim6 commented:
I sincerely apologize my Black formatting attempt accidentally reformatted many unrelated files. I'm cleaning this up now to isolate only the intended changes.

quim6 committed:
accelerated motif search (clean version)

quim6 committed:
accelerated motif search (clean version)

quim6 committed:
Argsument -> Argument typo fixed on search_algorithms.c

quim6 committed:
Argsument -> Argument typo fixed on search_algorithms.c

mdehoon committed:
Use 3.13.5 instead of 3.13

mdehoon committed:
Also use 3.13.5 when testing on Windows

mdehoon committed:
follow recommended style for PyTypeObject (#5014)

* follow recommended style for PyTypeObject

* update

---------

Co-authored-by: Michiel Jan Laurens de Hoon <mdehoon@Michiels-MacBook-Air.local>

mdehoon committed:
Style change only (#5019)

Co-authored-by: Michiel Jan Laurens de Hoon <mdehoon@Michiels-MacBook-Air.local>

mdehoon committed:
Reset to 3.13

3.13.5 is now cached on the runners, so specifying 3.13 will resolve to that version, and we don't need to hardcode it here any more.

JoaoRodrigues committed:
Update ci.yml and change windows-2019 to windows-2022 (#5024)

Closes #5023

dimi1729 committed:
Add automatic alignment generation to StructureAlignment

pre-commit-ci[bot] committed:
[pre-commit.ci] autoupdate

updates:
- [github.com/astral-sh/ruff-pre-commit: v0.9.7 → v0.12.2](https://github.com/astral-sh/ruff-pre-commit/compare/v0.9.7...v0.12.2)
- [github.com/psf/black: 24.10.0 → 25.1.0](https://github.com/psf/black/compare/24.10.0...25.1.0)
- [github.com/PyCQA/flake8: 7.1.2 → 7.3.0](https://github.com/PyCQA/flake8/compare/7.1.2...7.3.0)
- [github.com/pre-commit/mirrors-mypy: v1.15.0 → v1.16.1](https://github.com/pre-commit/mirrors-mypy/compare/v1.15.0...v1.16.1)
- [github.com/rstcheck/rstcheck: v6.2.4 → v6.2.5](https://github.com/rstcheck/rstcheck/compare/v6.2.4...v6.2.5)
- [github.com/PyCQA/doc8: v1.1.2 → v2.0.0](https://github.com/PyCQA/doc8/compare/v1.1.2...v2.0.0)

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

peterjc committed:
Drop trailing whitespace in some docstrings

Used black 25.1.0 which does this in docstrings too which is bad.

peterjc committed:
Stay at black 24.10.0 for trailing whitespace in doctests

mdehoon committed:
Extend the .counts method of an Alignment (#5011)

* update

* update

* update

* update

* update

* update

* update

* update

* update

* update

* update

* update

* update

* update

* update

* update

* update

* update

* update

* update

* update

* update

* update

* update

* update

* update

* update

* update

* update

* update

* update

* update

* update

* update

* update

* update

* update

* update

* update

* update

* update

* update

* update

* update

* update

* update

* update

* update

* update

* update

* update

* update

* update

* update

* update

* update

* update

* update

* update

* update

* update

* update

* update

* update

* update

* update

* update

* update

* update

* update

* update

* update

* update

* updat

* update

* update

* update

* update

* update

* update

* update

* update

* update

* update

* update

* update

* halfway finished

* update

* update

* update

* update

* update

* update

* update

* update

* update

* update

* update

* update

* update

* update

* update

* update

* update

* update

* update

* update

* update

* update

* update

* update

* update

* update

* pdate

* update

* update

* update

* update

* update

* update

* update

* update

* update

* update

* update

* update

* update

* update

* update

* replace precompiler #defines by inline functions

* update

* update

* add tests

* update

* update

* update

* update

* update

* update

* update

* update

* documentation

* update

* avoid stpcpy

* pointer printing

* compiler warning

* testing without codonalign

* testing without codonalign and pairwisealigner

* compiler warning

* adding codonalign and pairwisealigner back in

* remove inline from check_indices

* add inline to check_indices

* update

* add test line 12287 test_pairwise_aligner.py

* update

* update

* update

* update

* change dtype from int32 to "i"

* all done

* testing

* testing

* testing

* testing

* done

* done

* fix how pointers are printed on Windows

* update

* update

* update

* update

* fix id printing on pypy

* style change only

* Use Py_uintptr_t instead of uintptr_t

* fix memory leak

* remove double semicolon

* check if GitHub actions are now pickup up Python version 3.13.5 without hardcoding it

---------

Co-authored-by: Michiel Jan Laurens de Hoon <mdehoon@Michiels-MacBook-Air.local>
Co-authored-by: Michiel de Hoon <mdehoon@tkx288.genome.gsc.riken.jp>
Co-authored-by: Michiel de Hoon <michiel.dehoon@riken.jp>

Commit:
Prevent side-effect in SeqRecord._per_letter_annotations

Fixes #5015

twalsh-ebi committed:
Set PAML class names in CamelCase

twalsh-ebi committed:
Update long docstrings in PAML class methods

peterjc committed:
Drop comment

twalsh-ebi committed:
Uppercase PAML software names except in run methods

twalsh-ebi committed:
Update YN00 docstrings

mdehoon committed:
declare local functions as static (#5037)

Co-authored-by: Michiel Jan Laurens de Hoon <mdehoon@Michiels-MacBook-Air.local>

dependabot[bot] committed:
Bump actions/checkout from 4 to 5

Bumps [actions/checkout](https://github.com/actions/checkout) from 4 to 5.
- [Release notes](https://github.com/actions/checkout/releases)
- [Changelog](https://github.com/actions/checkout/blob/main/CHANGELOG.md)
- [Commits](https://github.com/actions/checkout/compare/v4...v5)

---
updated-dependencies:
- dependency-name: actions/checkout
  dependency-version: '5'
  dependency-type: direct:production
  update-type: version-update:semver-major
...

Signed-off-by: dependabot[bot] <support@github.com>

mdehoon committed:
typo (#5040)

Co-authored-by: Michiel de Hoon <michiel.dehoon@riken.jp>

amitschang committed:
pairwisealigner pickle fix (#5041)

* pairwisealigner pickling fix

* pr feedback updates

* pr updates - use almost equal on epsilon

OWissett committed:
Fixes formatting for extremely high B-factor values in PDB files (#5047)

RachelStern20 committed:
Align: add substitution matrix support to Alignment.format() (#5049)

* Align: add optional substitution matrix/aligner argument to format()
- Extended Alignment.format() to accept an optional argument:
  * substitution matrix (e.g. from Bio.Align.substitution_matrices)
  * PairwiseAligner object (uses its substitution matrix)
- Preserves compatibility with existing writer-specific arguments
  (e.g. "vulgar" for exonerate, metadata for mauve).
- Internal logic now distinguishes between substitution matrices
  (used in pretty-printing) and writer arguments (passed to
  AlignmentWriter).
- All alignment format tests pass (58/58).

* Align: add substitution matrix argument to helper format functions

- Updated _format_pretty, _format_unicode, _format_generalized
  to accept optional substitution matrix argument.
- Passed matrix from Alignment.format() into these helpers.
- Updated docstrings accordingly.
- All alignment format tests pass (58/58).

* Align.format: support substitution matrices in helper functions; add tests

- Updated helper functions in Align.format (generalized, pretty, html, etc.)
  to correctly handle substitution_matrix arguments.
- Added new test file `Tests/test_format_matrix.py` to cover:
  * identity matches (|), positive mismatches (:), negative mismatches (.),
    and gaps (-) using substitution matrices.
  * verified behavior with both NUC.4.4 and a custom BLASTN-like matrix.
- Confirmed tests pass with pytest.

Resolves #5043

* Add Rachel Stern to contributors list and NEWS (first contribution)

* [pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

* Fix style issues in test_format_matrix and Align -__init__

* Rename test_format_matrix.py to test_format_matrix_unittest.py for AppVeyor test discovery

* Move format matrix test to unittest style

* Refactor scoring argument handling in counts() and format()

- Renamed the optional argument in counts() from `argument` to `scoring` for
  clarity and consistency.
- Updated format() to accept the `scoring` parameter via **kwargs instead of
  a dedicated argument.
- Removed the explicit writer argument from format(), since it is no longer
  needed with kwargs-based handling.
- Adjusted the docstring of format() accordingly to reflect the new usage.

* Update test_format_matrix_unittest.py:
- Use explicit assertEqual for pattern strings instead of only assertIn/NotIn, following reviewer feedback.
- Simplify alignment selection by indexing [0] instead of next(iter(...)).

* Remove _pattern_from_pretty, assert full alignment format instead

- Replaced usage of `_pattern_from_pretty` helper with direct comparison
  of `aln.format("", scoring=M)` output in all relevant tests.
- This makes the tests simpler and closer to the actual API usage.
- Removed unused `_pattern_from_pretty` function.

* docs: document new `scoring` argument in Alignment.format tutorial

- Added explanation of the optional `scoring` argument to
  `Alignment.format()` in Doc/Tutorial/chapter_align.rst.
- Included example using PairwiseAligner and a substitution matrix
  (NUC.4.4) to illustrate how the pattern line reflects scores.
- Clarifies how symbols (|, :, ., -) correspond to matches,
  positive/negative mismatches, and gaps.

* docs: add cont-doctest and update scoring example with '.' mismatch

* docs: fix scoring example with gap penalties to show ':' and '.' in doctest

* docs: add <BLANKLINE> to doctest output in scoring example

* doc: add 8 in the middle line of the format

* doc: Arranging the print format

* Rearrange the print format

---------

Co-authored-by: pre-commit-ci[bot] <66853113+pre-commit-ci[bot]@users.noreply.github.com>
Co-authored-by: mdehoon <mjldehoon@yahoo.com>

mdehoon committed:
documentation tweaks (#5055)

* documentation tweaks

* update

---------

Co-authored-by: Michiel de Hoon <mdehoon@tkx288.genome.gsc.riken.jp>

dependabot[bot] committed:
Bump actions/setup-python from 5 to 6

Bumps [actions/setup-python](https://github.com/actions/setup-python) from 5 to 6.
- [Release notes](https://github.com/actions/setup-python/releases)
- [Commits](https://github.com/actions/setup-python/compare/v5...v6)

---
updated-dependencies:
- dependency-name: actions/setup-python
  dependency-version: '6'
  dependency-type: direct:production
  update-type: version-update:semver-major
...

Signed-off-by: dependabot[bot] <support@github.com>

dependabot[bot] committed:
Bump tj-actions/changed-files from 46 to 47

Bumps [tj-actions/changed-files](https://github.com/tj-actions/changed-files) from 46 to 47.
- [Release notes](https://github.com/tj-actions/changed-files/releases)
- [Changelog](https://github.com/tj-actions/changed-files/blob/main/HISTORY.md)
- [Commits](https://github.com/tj-actions/changed-files/compare/v46...v47)

---
updated-dependencies:
- dependency-name: tj-actions/changed-files
  dependency-version: '47'
  dependency-type: direct:production
  update-type: version-update:semver-major
...

Signed-off-by: dependabot[bot] <support@github.com>

emollier committed:
Delete Tests/BWA/ test data.

This directory stored data files for the test_BWA_tool.py.  The latter
test was removed following Bio.Application removal in issue #4906.  No
other files in the source tree look to be referring to these data,
therefore it should be safe to remove the whole set.

Closes: #5064
Signed-Off-By: Étienne Mollier <emollier@debian.org>

mdehoon committed:
Aligner default arguments (#5029)

* update

* update

* update

* update

* update

* update

* update

* update

* update

* update

* update

* update

* update

* update

* update

* update

* update

* update

* update

* update

* update

* update

* doctests_fixed

* update

* update

* documentation

* add warning

* add a note to the DEPRECATED file

* adding a NEWS entry

* remove stray comments

---------

Co-authored-by: Michiel de Hoon <mdehoon@lacg01.local>
Co-authored-by: Michiel de Hoon <mdehoon@tkx380.genome.gsc.riken.jp>
Co-authored-by: Michiel Jan Laurens de Hoon <mdehoon@Michiels-MacBook-Air.local>
Co-authored-by: Michiel de Hoon <mdehoon@tkx288.genome.gsc.riken.jp>

Commit:
removing deprecated code

Commit:
updating the DEPRECATED.rst file

Commit:
removing tests for removed modules

Mikel-MG committed:
Fix F824 in Bio.PDB.vectors (#5067)

Remove unnecessary global statements

mdehoon committed:
deprecating the SummaryInfo class in Bio.Align.AlignInfo (#5069)

gouttegd committed:
SnapGene: Filter out partial binding sites.

The "primers" packet within a SnapGene file may contain, for every
primers, additional partial binding sites in addition to the full
binding site.

We use the minimum values described in the `HybridizationParams` element
to filter out any binding site that:

* has a matching length shorter than the minimum accepted matching
  length, or
* has a melting temperature lower than the minimum accepted melting
  temperature.

closes #5053

manulera committed:
added test and docstring for #5053

chenyenchung committed:
Handle possible non-existent attributes in clusterbuster.py (#5074)

* Handle possible non-existent attrs in clusterbuster.py

In commit bc0d58b, support was added for gap and weight paramters, but
since these two attributes are not guaranteed to exist by parse and read
methods, an AttributeError could be raised for non-clusterbuster motifs.

An attribute check was added to allow motifs lacking gap and weight to
be formatted properly.

Co-authored-by: Ee Shan Liau Shirley <69098782+shirleyliau@users.noreply.github.com>

* Add names to CONTRIB.rst

* Use try-except to handle missing attributes

Co-authored-by: Ee Shan Liau Shirley <69098782+shirleyliau@users.noreply.github.com>

* Avoid CB output when an attribute is None

In the previous commit, the skipping mechanism was switched to
try-except blocks, and the cases when an attribute is None are
no longer dealt with. This patch re-introduced `if m.weight`
to prevent empty export fields in such cases.

Co-authored-by: Ee Shan Liau Shirley <69098782+shirleyliau@users.noreply.github.com>

---------

Co-authored-by: Ee Shan Liau Shirley <69098782+shirleyliau@users.noreply.github.com>

kamurani committed:
Make `PDB.Chain.__getitem__` more lenient for indexing with `int`-like values  (#5018)

Includes residue indexing test case for `np.int64` condition

pre-commit-ci[bot] committed:
[pre-commit.ci] autoupdate

updates:
- [github.com/pre-commit/pre-commit-hooks: v5.0.0 → v6.0.0](https://github.com/pre-commit/pre-commit-hooks/compare/v5.0.0...v6.0.0)
- [github.com/astral-sh/ruff-pre-commit: v0.12.2 → v0.13.3](https://github.com/astral-sh/ruff-pre-commit/compare/v0.12.2...v0.13.3)
- https://github.com/psf/black → https://github.com/psf/black-pre-commit-mirror
- [github.com/psf/black-pre-commit-mirror: 24.10.0 → 25.9.0](https://github.com/psf/black-pre-commit-mirror/compare/24.10.0...25.9.0)
- [github.com/pre-commit/mirrors-mypy: v1.16.1 → v1.18.2](https://github.com/pre-commit/mirrors-mypy/compare/v1.16.1...v1.18.2)
- [github.com/asottile/blacken-docs: 1.19.1 → 1.20.0](https://github.com/asottile/blacken-docs/compare/1.19.1...1.20.0)

peterjc committed:
Stay at black 24.10.0

peterjc committed:
B017 - avoid blind exception

peterjc committed:
Stay on mypy 1.16.1

iosonofabio committed:
Add paragraph about iplotx in Phylo tutorial (#5028)

Add mention in NEWS.rst

peterjc committed:
Test on Python 3.10 ~ 3.14

peterjc committed:
Sort names

[ci skip]

peterjc committed:
Fix markup, update & sort names

peterjc committed:
Remove unused or read only globals (flake8 F824)

peterjc committed:
Replaced deprecated ruff linter ID

peterjc committed:
ruff v0.11.10 format then black 25.9.0

Note ruff still ignores the fmt off/on comments,
and like recent black removes trailing whitespace
in doctest output.

peterjc committed:
Silence warning for translating when not a multiple of 3

Cassiebastress committed:
From pairwise alignments (#5051)

* Add classmethod from_pairwise_alignments to MultipleSeqAlignment

* Wrote tests for from_pairwise_alignments in test_align.py (all pass)

* manu feedback

* Writing more tests, anddiscovering edge case behavior (see commented out test)

* Added metadata preservation functionality to from_pairwise_alignments() and added more comprehensive tests

* Added two ways to preserve metadata using SeqIO functions, must decide which is better.

* simplify parsing and restoring id and description

* Added docstrings with examples to from_pairwise_alignments

* Added my mame to CONTRIB.rst

* Work in progress: trying to build correct coordinates array in order to move method to the Alignment class

* Debugged from_pairwise_alignments, all tests pass

* Fixed docstrings tests, setup.py test --offline now passes

* 1st step towards working with coordinates

* modify tests & remove unused code

* handle undefined sequences

* Added tests and made changes to accommodate input Alignments of more than two sequences

* Writing tests, adding sequence padding to method

* replace deprecated method

* improve docstring

* address revision comments

* Add from_alignments_with_same_reference to NEWS.rst and contributors list

* Remove unused or read only globals (flake8 F824)

* Replaced deprecated ruff linter ID

* Changed logic to use coordinates, note that the method ONLY works for pairwise alignment input now

* adapt to support merging of alignments with more than 1 query

* Added commenting

* Changed pwas to alignments

* Fix type

* Added documentation for from_alignments_with_same_reference to Doc/Tutorial/chapter_align.rst

* Typo fixes in documentation

* [pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

* fix style

* fix rst hook

* remove unused imports

---------

Co-authored-by: Manuel Lera-Ramirez <manulera14@gmail.com>
Co-authored-by: Peter J. A. Cock <p.j.a.cock@googlemail.com>
Co-authored-by: pre-commit-ci[bot] <66853113+pre-commit-ci[bot]@users.noreply.github.com>

corneliusroemer committed:
doc: Fix typo `Thie` -> `This` in fasta warning

mikeph52 committed:
silence test_GenBank.py invalid date warnings in test suite

mikeph52 committed:
Add Mike Philippakis to contributor list

peterjc committed:
Fix trailing whitespace with black

mikeph52 committed:
Fix issues that prevented black

peterjc committed:
Silence PDB warnings in tests (add fix name mix-up)

peterjc committed:
Update license metadata as per Core Metadata specification v2.4 (August 2024)

See https://packaging.python.org/en/latest/specifications/core-metadata/

This solves a deprecation warning:

SetuptoolsDeprecationWarning: License classifiers are deprecated.
!!

        ********************************************************************************
        Please consider removing the following classifiers in favor of a SPDX license expression:

        License :: Freely Distributable

        See https://packaging.python.org/en/latest/guides/writing-pyproject-toml/#license for details.
        ********************************************************************************

!!

peterjc committed:
Call this Biopython 1.86

peterjc committed:
Post-release version bump

[ci skip]

peterjc committed:
Post-release version bump

[ci skip]

peterjc committed:
Revert "Post-release version bump"

This reverts commit 708b127690347f47a1efacaf75d01ea005b8a7be.

peterjc committed:
Revert to old license metadata

PyPI didn't like License-Expression: LicenseRef-Biopython-License-Agreement

[ci skip]

peterjc committed:
Fully revert to old license metadata

PyPI didn't like 'License-File: LICENSE.rst' either.

[ci skip]

mdehoon committed:
Remove unused import (#5093)

mdehoon committed:
Remove unused imports in test_AlignIO.py

peterjc committed:
Update HTML tutorial link (old one worked via a redirect)

peterjc committed:
Update FAQ on latest version of the docs

peterjc committed:
Remove reference to PDF Tutorial

See https://github.com/biopython/biopython.github.io/issues/213

peterjc committed:
Bump year in license (overdue)

[ci skip]

timjden committed:
Update ScanProsite for new XML format from ScanProsite API (#5099)

* add helper function to check for valid XML
* adjust element paths to include new scanprosite_response wrapper element
* add name to contributors and news
* put the expasy url in a constant so it can be referenced by the tests
* validate that the output param is set to xml and raise NotImplemented if not
* scanprosite tests (requires internet)
* an offline test for ScanProsite.read()

timjden committed:
fix error message raised when the target is not found in the tree

timjden committed:
add missing space in the format string

timjden committed:
test for fixed format string

hartwork committed:
Make pre-commit CI more helpful

.. by showing the diff on failure

dependabot[bot] committed:
Bump actions/checkout from 5 to 6

Bumps [actions/checkout](https://github.com/actions/checkout) from 5 to 6.
- [Release notes](https://github.com/actions/checkout/releases)
- [Changelog](https://github.com/actions/checkout/blob/main/CHANGELOG.md)
- [Commits](https://github.com/actions/checkout/compare/v5...v6)

---
updated-dependencies:
- dependency-name: actions/checkout
  dependency-version: '6'
  dependency-type: direct:production
  update-type: version-update:semver-major
...

Signed-off-by: dependabot[bot] <support@github.com>

hartwork committed:
Remove Python 3.8 support leftovers

peterjc committed:
Thank Ziyan Rao in NEWS.rst and CONTRIB.rst (#5111)

See #5098.

[ci skip]

Mikel-MG committed:
Fix typo/bug in biopython.Bio.PDB.vectors vector_to_axis (#5045)

Add testing for vector_to_axis

hartwork committed:
README.rst: Mass-replace http:// links with https://

Command used:
> sed 's,http://,https://,g' -i.bak README.rst

timjden committed:
use an existing pdb file for the test

manulera committed:
Fix parsing of snapgene features with linebreaks in qualifier values (#5117)

Closes #5116 by mapping any embedded new line to space when parsed.

dependabot[bot] committed:
Bump actions/cache from 4 to 5

Bumps [actions/cache](https://github.com/actions/cache) from 4 to 5.
- [Release notes](https://github.com/actions/cache/releases)
- [Changelog](https://github.com/actions/cache/blob/main/RELEASES.md)
- [Commits](https://github.com/actions/cache/compare/v4...v5)

---
updated-dependencies:
- dependency-name: actions/cache
  dependency-version: '5'
  dependency-type: direct:production
  update-type: version-update:semver-major
...

Signed-off-by: dependabot[bot] <support@github.com>

suyash469 committed:
DOC: Clarify Visual Studio Build Tools workload requirement for Windows (#5118)

* docs: clarify Visual Studio Build Tools requirements for Windows

* Fix README formatting per review

* Fix run on text; drop old term Mac OS

---------

Co-authored-by: Peter Cock <p.j.a.cock@googlemail.com>

beneopp committed:
Issue 5122 bug fix (#5126)

* fixed bug mentioned in issue #5122 by skipping any lines in chain file starting with '#'

* Revert "fixed bug mentioned in issue #5122 by skipping any lines in chain file starting with '#'"

This reverts commit b79ac20163b95b046707b85e3fd5cd6148808c2d.

* fixed Issue #5122

* Add Ben Oppenheimer to CONTRIB.rst

pre-commit-ci[bot] committed:
[pre-commit.ci] autoupdate

updates:
- [github.com/astral-sh/ruff-pre-commit: v0.13.3 → v0.14.10](https://github.com/astral-sh/ruff-pre-commit/compare/v0.13.3...v0.14.10)
- [github.com/psf/black-pre-commit-mirror: 24.10.0 → 25.12.0](https://github.com/psf/black-pre-commit-mirror/compare/24.10.0...25.12.0)
- [github.com/pre-commit/mirrors-mypy: v1.16.1 → v1.19.1](https://github.com/pre-commit/mirrors-mypy/compare/v1.16.1...v1.19.1)

kirandharmar867-maker committed:
Run PSEA in a temporary directory to avoid leftover files

csbiy committed:
Silence PDBConstructionWarning in test_PDB_StructureAlignment

Wrap PDB structure parsing calls with warnings.catch_warnings() to
suppress PDBConstructionWarning messages during test execution.

This follows the same pattern used in other test files such as
test_PDB_MMCIFParser.py, test_mmtf_online.py, and test_RCSBFormats.py.

Fixes #1877

peterjc committed:
Ignore B905

peterjc committed:
Ignore B905 in ruff command line too

peterjc committed:
Stay on mypy 1.16.1 and older black

peterjc committed:
Replace Union and Optional with Python 3.10+ typing

kirandharmar867-maker committed:
Format Bio/PDB/PSEA.py with black

csbiy committed:
Replace print statements with warnings/stderr in NCBIXML (#5140)

- Replace debug print statements with sys.stderr.write() for debug mode
- Replace CREATE_VIEW print with warnings.warn() using BiopythonParserWarning
- Add test case for CREATE_VIEW warning

Fixes #4262

kirandharmar867-maker committed:
Add test for run_psea using a temporary directory

mdehoon committed:
Remove code that was deprecated in 1.85 (#5144)

mdehoon committed:
Avoid http links in Bio.Entrez (#5114)

This also validates DTD and XSD URLs to be HTTPS and from
"www.ncbi.nlm.nih.gov", "dtd.nlm.nih.gov", or "eutils.ncbi.nlm.nih.gov".

xelandernt committed:
Migrate to pyproject.toml

peterjc committed:
Fix unstable doctest (key order random)

xelandernt committed:
Correctly clean up files generated by the Phylo tutorial.

peterjc committed:
Use more typical double indexing with numpy arrays

peterjc committed:
Drop older commented code lines in PDB vector code

peterjc committed:
Fix vector test for NumPy 2.4

manulera committed:
closes #5162

piecole committed:
Added PERMISSIVE argument to MMCIFParser as in PDBParser. (#5134)

This is to allow for problems in PDB file with e.g. duplicate atoms in a residue.

Kept PERMISSIVE default False to keep behaviour consistent with previous versions.

peterjc committed:
Using np.sin or np.cos was generating return values which were not plain float.

hartwork committed:
Fix `Entrez.Parser.DataHandler` URL security check continuity (related to #5109 and #5114) (#5148)

* Bio/Entrez/Parser.py: Do not stop checking URLs after first success

Previously, one passed check from DataHandler.verify_security
would disabled checks for any consecutive requests to URLs,
independent of URL scheme and target host.

* Bio/Entrez/Parser.py: Rename flag .secure to .bypass_url_security

.. to better communicate what is doing

* Bio/Entrez/Parser.py: Fix DataHandler.verify_security docstring

* Bio/Entrez/Parser.py: Improve DataHandler.verify_security error message

* Bio/Entrez/Parser.py: Add host jats.nlm.nih.gov to trusted URL hosts

This URL is used by two existing tests:
https://jats.nlm.nih.gov/archiving/1.3/JATS-archivearticle1-3-mathml3.dtd

* Tests/test_Entrez_parser.py: Cover DataHandler.verify_security continuity

* CONTRIB.rst|NEWS.rst: Add myself as a contributor for Biopython 1.87

* Bio/Entrez/Parser.py: Trust all *.nlm.nih.gov domains

.. to allow changes on subdomain level in the future

Co-authored-by: Peter Cock <p.j.a.cock@googlemail.com>

* Bio/Entrez/Parser.py: Trust external references from NCBI files

General idea by Michiel de Hoon.

* Bio/Entrez/Parser.py: Improve DataHandler.verify_security docstring

Co-authored-by: Peter Cock <p.j.a.cock@googlemail.com>

---------

Co-authored-by: Peter Cock <p.j.a.cock@googlemail.com>

peterjc committed:
Stress importance of the PR template

There have been several PRs recently not following the template - some using generative AI even, if not done by a software agent.

In either case, this adds to the burden of review (we don't want anything improperly licensed to get merged).

Try to make our expectations clearer, and make it easier to just reject unwanted PRs.

dependabot[bot] committed:
Bump geekyeggo/delete-artifact from 5 to 6

Bumps [geekyeggo/delete-artifact](https://github.com/geekyeggo/delete-artifact) from 5 to 6.
- [Release notes](https://github.com/geekyeggo/delete-artifact/releases)
- [Changelog](https://github.com/GeekyEggo/delete-artifact/blob/main/CHANGELOG.md)
- [Commits](https://github.com/geekyeggo/delete-artifact/compare/v5...v6)

---
updated-dependencies:
- dependency-name: geekyeggo/delete-artifact
  dependency-version: '6'
  dependency-type: direct:production
  update-type: version-update:semver-major
...

Signed-off-by: dependabot[bot] <support@github.com>

peterjc committed:
Biopython 1.87

dependabot[bot] committed:
Bump codecov/codecov-action from 5 to 6

Bumps [codecov/codecov-action](https://github.com/codecov/codecov-action) from 5 to 6.
- [Release notes](https://github.com/codecov/codecov-action/releases)
- [Changelog](https://github.com/codecov/codecov-action/blob/main/CHANGELOG.md)
- [Commits](https://github.com/codecov/codecov-action/compare/v5...v6)

---
updated-dependencies:
- dependency-name: codecov/codecov-action
  dependency-version: '6'
  dependency-type: direct:production
  update-type: version-update:semver-major
...

Signed-off-by: dependabot[bot] <support@github.com>

peterjc committed:
Automatically update the docs symlink for new releases

Also ran shfmt -i 4 -w .github/deploy_docs.sh
(whitespace and minor style changes).

peterjc committed:
Automatically update the docs symlink for new releases - retry

peterjc committed:
Automatically update the docs symlink for new releases - Take 3

peterjc committed:
Automatically update the docs symlink for new releases - Take 4

Forgot about the trailing slash.

peterjc committed:
Automatically update the docs symlink for new releases - Take 5

peterjc committed:
Automatically update the docs symlink for new releases - Take 6

peterjc committed:
Automatically update the docs symlink for new releases - Take 7

A shame this script only runs on the default branch...

peterjc committed:
Post Biopython 1.87 release version bump

[ci skip]

peterjc committed:
Remove stray file

Added in error in #5071

peterjc committed:
Declare Python 3.10 support deprecated.

No obvious way to add a warning with pyproject.toml like
we used to do with setup.py?

quim6 committed:
deleted numpy as a dependency

quim6 committed:
deleted numpy as a dependency

quim6 committed:
Missed deletion in _searchmodule.c

quim6 committed:
Missed deletion in _searchmodule.c

peterjc commented on Tests/test_motifs.py:
```diff
@@ -5026,7 +5026,10 @@ def test_TFoutput(self):
+    import os
+
+    this_dir = os.path.dirname(__file__)
+    with open(os.path.join(this_dir, "motifs", "SRF.pfm")) as stream:
```
Revert to using `with open("motifs/SRF.pfm") as stream:` please, all the tests assume they are run from the tests sub-directory as the alternative is too verbose and would quickly get inconsistent.

[The doctests and Tutorial tests are a special case]

quim6 replied:
Thank you, I will change it right away.

peterjc committed:
Fix issue with GenBank debug print (#5190)

Now follows style used in EMBL parser earlier in the file.

Cross reference #5189, this was spotted by an updated ruff.

quim6 committed:
Revert to relative path from tests directory

quim6 committed:
Revert to relative path from tests directory

quim6 committed:
new setup.py

peterjc commented on setup.py:
```diff
@@ -195,6 +194,14 @@ def can_import(module_name):
+            "Bio/motifs/search_algorithms.c",
+            "Bio/motifs/darray.c",
+        ],
+    ),
```
You'll have to deal with the move from `setup.py` to `pyproject.toml` done in #5142.

If you want to keep the commit history, can you do a git rebase? Otherwise, just resolve the conflicts with a merge, and if this PR is accepted it will probably have to be a squash-commit (all your changes as one commit to our default branch).

quim6 replied:
Thank you Peter. I did the git rebase and adapted the changes to move from setup.py to pyproject.toml.

Please let me know if anything else needs to be adjusted, or if I should reach out to any of the reviewers for further feedback.

quim6 committed:
Moved from setup.py to pyproject.toml

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

ivanerill commented:
> Do you know what have you done to have GitHub say your PR changes over 200 files?
> 
> If you are familiar with `git rebase origin/master` could you try that? If you're not familiar with `git rebase` that might be rather tricky... so safer not to touch it.

@peterjc , @quim6's code was developed on a one-year-old version of biopython, so that might be causing issues. Would it be better if Quim simply forked anew the current version of biopython, introduced his code changes and retested, then submitted a new pull-request?

peterjc commented:
@ivanerill I think the `git rebase` would be a simpler way to achieve that - without needing a new PR. However one could also start with a new branch from the current default branch, and apply the changes with `git cherry-pick` or manually... that might seem easier. I'd look at the merge conflicts to get a feel for this (if just minor conflicts I'd do the rebase route, complex conflicts becomes less clear cut).

Note the rebase should be to the main repository's `master` branch, not any out-of-date local or fork's `master` branch - that might explain why the rebase two weeks ago didn't clean up the history.

quim6 commented:
@peterjc Hello Peter, I think I'll go with creating a new branch from the current master and open a new PR instead. That seems like the simplest approach for me, since I'm not familiar with git rebase.
